### ISSUE TITLE
Character Encoding Error on UTF-8 Encoded Source File with U+0441

### ISSUE THREAD

**Issue by kuchungmsft** [david-a-wheeler/flawfinder#68]
## Encoding Error on Windows 10
```
git clone https://github.com/Kitware/CMake.git

git clone https://github.com/david-a-wheeler/flawfinder.git

python3 flawfinder\flawfinder.py CMake\Source\CTest\cmCTestBuildHandler.cxx
```
![image](https://user-images.githubusercontent.com/105310954/169356614-0d893d75-c5e0-463b-8870-39313db64f9c.png)

![image](https://user-images.githubusercontent.com/105310954/169357935-4fc54b09-5a03-4efb-b355-71d48c9b92e9.png)

## No Encoding Error on WSL

![image](https://user-images.githubusercontent.com/105310954/169357564-288509bb-6451-4851-bc40-3379b99d8fa6.png)

## File Content
![image](https://user-images.githubusercontent.com/105310954/169357137-f7e432f2-e39d-4cc9-b3ee-6bf36a1dafe7.png)

**Comment by david-a-wheeler** [comment]
This looks like the text isn't actually UTF-8 in the file being analyzed. Have you verified that the file being examined actually complies with UTF-8?

If it doesn't comply with UTF-8 (seems likely), see the documentation on various options. Sadly, Python3 doesn't provide good tools for handling non-UTF-8 text files.

**Comment by kuchungmsft** [comment]
Notepad++ thinks the file is UTF-8.
![image](https://user-images.githubusercontent.com/105310954/169375768-f75c2d89-e74c-4303-8c39-c9a77554728a.png)

VS Code thinks the file is UTF-8.
![image](https://user-images.githubusercontent.com/105310954/169375561-485d4ea9-2a4e-4cbd-87d3-dbb20eb72891.png)

Notepad thinks the file is UTF-8.
![image](https://user-images.githubusercontent.com/105310954/169376299-d28a1ede-0479-4b48-b296-4809eb89e9b1.png)


I think the file does comply with UTF-8.

**Comment by david-a-wheeler** [comment]
Also: if the character is just a *literal* 0x81 byte, that is not valid UTF-8.

**Comment by kuchungmsft** [comment]
My bad, it's character U+0441, I have updated the title.
![image](https://user-images.githubusercontent.com/105310954/169400547-b02a5b80-39ec-4238-9246-78a424e34c0f.png)

Tried "iconv", no difference between original file and the converted one.
![image](https://user-images.githubusercontent.com/105310954/169403990-826be8f1-a222-41d2-bccb-0f1fb2019e73.png)

**Comment by david-a-wheeler** [comment]
Huh. That doesn't make sense to me at all. The sequence 0xd1 0x81 seems like perfectly fine UTF-8 to me, it shouldn't give you that error message.

So we agree it shouldn't happen. But clearly it's happening anyway :-).

Can you send me a URL for a mishandled file so I can just use curl/wget to get it? Ideally make the test file as small as possible while still causing the problem. I want to reproduce the problem with the smallest possible failing test. If I can reproduce it, I should be able to fix it, or at least explain it & suggest a workaround.

**Comment by david-a-wheeler** [comment]
Oh, weird thought: This is on Windows. Is it *possible* it's actually being stored as UTF-16? I doubt that's what is going on, but I'm grasping at straws and maybe this is the straw I needed :-).

**Comment by kuchungmsft** [comment]
Here is the link to that file: https://github.com/Kitware/CMake/blob/master/Source/CTest/cmCTestBuildHandler.cxx

I am not sure how to check if it's stored as UTF-16, I mean it's just a plain text file, I don't see any header when viewing it in hex.
![image](https://user-images.githubusercontent.com/105310954/169419489-081ce62b-c5d2-406c-9747-48e025f12a37.png)

**Comment by kuchungmsft** [comment]
This character is what you need to reproduce the issue.

![image](https://user-images.githubusercontent.com/105310954/169419661-5d325874-15c5-491c-bb7b-d32eebafaefb.png)

**Comment by david-a-wheeler** [comment]
You posted an image showing the file. However, I need the file contents itself. Can you post it somewhere (ideally shortened) & share the URL to it? A small snippet would be best for my purposes.

**Comment by david-a-wheeler** [comment]
Oh, whups, you *did* provide a link. Thank you.

I ran it on MacOS and it worked just fine. Below is the output.

Ugh, it seems to be a Windows 10 specific thing. I don't have any of those platforms.
I want to fix it, but I have to be able to reproduce it. Any ideas?

~~~~
python3 ./flawfinder.py cmCTestBuildHandler.cxx 
Flawfinder version 2.0.19, (C) 2001-2019 David A. Wheeler.
Number of rules (primarily dangerous function names) in C/C++ ruleset: 222
Examining cmCTestBuildHandler.cxx

FINAL RESULTS:

cmCTestBuildHandler.cxx:6204:  [2] (misc) open:
  Check when opening files - can an attacker redirect it (via symlinks),
  force the opening of special file type (e.g., device files), move things
  around to create a race condition, control its ancestors, or change its
  contents? (CWE-362).

ANALYSIS SUMMARY:

Hits = 1
Lines analyzed = 6240 in approximately 0.28 seconds (22118 lines/second)
Physical Source Lines of Code (SLOC) = 364
Hits@level = [0]   0 [1]   0 [2]   1 [3]   0 [4]   0 [5]   0
Hits@level+ = [0+]   1 [1+]   1 [2+]   1 [3+]   0 [4+]   0 [5+]   0
Hits/KSLOC@level+ = [0+] 2.74725 [1+] 2.74725 [2+] 2.74725 [3+]   0 [4+]   0 [5+]   0
Minimum risk level = 1

Not every hit is necessarily a security vulnerability.
You can inhibit a report by adding a comment in this form:
// flawfinder: ignore
Make *sure* it's a false positive!
You can use the option --neverignore to show these.

There may be other security vulnerabilities; review your code!
See 'Secure Programming HOWTO'
(https://dwheeler.com/secure-programs) for more information.
~~~~

**Comment by kuchungmsft** [comment]
Yeah, I couldn't repro it using WSL Ubuntu. Looks like this issue is not easy to tackle, I wonder if detecting encoding using chardet before opening the file an acceptable solution?

https://stackoverflow.com/questions/36303919/python-3-0-open-default-encoding

https://peps.python.org/pep-0597/

https://chardet.readthedocs.io/en/latest/usage.html#example-using-the-detect-function

**Comment by david-a-wheeler** [comment]
Hmm, it appears you're trying to process UTF-8 files, but the Windows default is *NOT* UTF-8, and that's the mismatch.

Try this:

python3 -X utf8 flawfinder.py ....

or set `PYTHONUTF8` to 1. In a shell do this:

~~~~
export PYTHONUTF8=1  # linux / macOS
set PYTHONUTF8=1  # windows
~~~~

.. .then run flawfinder.

**Comment by kuchungmsft** [comment]
That CMake repo has [Tests/RunCMake/CommandLine/cmake_depends/test_UTF-16LE.h](https://github.com/Kitware/CMake/blob/master/Tests/RunCMake/CommandLine/cmake_depends/test_UTF-16LE.h) in UTF-16 encoding, if I force it by `set PYTHONUTF8=1`, I get encoding error on that file.
![image](https://user-images.githubusercontent.com/105310954/169615579-af224c94-dfae-44b0-af77-2008d752cb08.png)

Is there a way to exclude certain folders that contain non-product code, i.e., in this case `Tests` folder.

**Comment by david-a-wheeler** [comment]
There isn't an `--exclude` option though that's not a bad idea. However, you can expressly list just the files and/or directories to scan, so just be more explicit about it.

However: can you tell me if PYTHONUTF8=1 resolves the problem with [cmCTestBuildHandler.cxx](https://github.com/Kitware/CMake/blob/master/Source/CTest/cmCTestBuildHandler.cxx) ? If it does, then we're at least making progress.

Flawfinder doesn't have a way of scanning different files with different encodings. Most software developers wouldn't want to do that. If you have to do that, I suggest making a *copy*, changing all the source files to some consistent encoding, and then analyzing them.

**Comment by kuchungmsft** [comment]
Yes, `PYTHONUTF8=1` resolves problem with [cmCTestBuildHandler.cxx](https://github.com/Kitware/CMake/blob/master/Source/CTest/cmCTestBuildHandler.cxx), thanks. The suggestion to make a copy and have a consistent encoding would not work for me because [test_UTF-16LE.h](https://github.com/Kitware/CMake/blob/master/Tests/RunCMake/CommandLine/cmake_depends/test_UTF-16LE.h) is meant to validate that CMake can handle UTF-16, just like compilers can handle inconsistent encoding of source files.

I guess I can workaround it by analyzing only the `Source` folder instead of entire repo. Thanks a lot for your help.

**Comment by imba-tjd** [comment]
On windows, the default encoding for `open()` is a thing called "ANSI", or MBCS.

However, I would suggest to always use utf-8, ignoring those who doesn't use utf-8 for their source code. Because the user of this software is programmer, not the end-user. I believe the number of programmer who uses utf-8 is greater than ansi, although they are using windows.

**Comment by david-a-wheeler** [comment]
> However, I would suggest to always use utf-8, ignoring those who doesn't use utf-8 for their source code.

The problem is that not all source code uses UTF-8. It's a reasonable default, certainly.

**Comment by david-a-wheeler** [comment]
I've made changes that I think will help on character encoding detection and reporting, as well as an `--exclude` option. I think that is enough to resolve this.
